### PR TITLE
Provide the ability to specify custom layouts from the config

### PR DESCRIPTION
**Pull request by mathias-lang-sociomantic**
Issue reported by @daniel-zullo-sociomantic :

> Some configureLoggers() were deprecated in v2.6.0 and the suggested method to use instead configureOldLoggers() does not deal with the template parameters FileLayout and ConsoleLayout.

This provides a more generic approach, less tied to the internals of this function (and that can be done at runtime).

CC @daniel-zullo-sociomantic @hatem-oraby-sociomantic for testing / sanity.

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
Note: A similar fix will need to be applied to v3.x.x (and the merge will conflict since it's touching deprecated code).

File: relnotes/makelayout.migration.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        ocean.util.log.Config.configureOldLoggers(config, m_config,
+            file_appender, use_insert_appender, makeLayout);
+    }
+    ```
+    Note that the previous example is kept short for simplicity. `makeLayout` gets the raw name,
+    and as such should make sure to make the comparison as loose as wanted (by lowercasing,
+    removing `"layout"` prefix/suffix, etc...).
+
+    In addition, the members `console_layout` and `file_layout` in the `Config` class
```
Comment (gavin-norman-sociomantic) [inline relnotes/makelayout.migration.md]: I wonder, would it be possible to leave the default values as an empty string and add an extra check that converts `""` -> `"simple"` or `"date"` in the point where the fields are referred to, in `configureLogger`? Would that remove the need for this migration instruction and make this a pure new feature?

Comment (mathias-lang-sociomantic) [reply relnotes/makelayout.migration.md]: I think I got mixed up, because this is v3.0.0 material. Removed it.

**Comment by mathias-lang-sociomantic** [comment]
Updated:
```diff
diff --git a/relnotes/makelayout.feature.md b/relnotes/makelayout.feature.md
index f92a097..dabb75b 100644
--- a/relnotes/makelayout.feature.md
+++ b/relnotes/makelayout.feature.md
@@ -1,6 +1,6 @@
 * `ocean.util.log.Config`
 
-    The `configureLogger`, `configureLoggers` overloads and `configureOldLoggers`
+    The `configureLogger` and `configureOldLoggers` functions
     have gained an additional `makeLayout` parameter.
     This parameter is used to instantiate a `Layout` from a name, allowing
     users to have custom layout usable from their config files.
@@ -35,9 +35,3 @@
     Note that the previous example is kept short for simplicity. `makeLayout` gets the raw name,
     and as such should make sure to make the comparison as loose as wanted (by lowercasing,
     removing `"layout"` prefix/suffix, etc...).
-
-    In addition, the members `console_layout` and `file_layout` in the `Config` class
-    (used to configure a logger instance) now default to `"simple"` and `"date"`, respectively.
-    This means that if any code relying on them being empty to instantiate the default values
-    (previously provided as `FileLayout` and `ConsoleLayout` template parameters) should now
-    rely on providing a cusom `makeLayout` parameter.
diff --git a/src/ocean/util/log/Config.d b/src/ocean/util/log/Config.d
index 18e9e58..ff5ec23 100644
--- a/src/ocean/util/log/Config.d
+++ b/src/ocean/util/log/Config.d
@@ -151,7 +151,7 @@ class Config
 
     ***************************************************************************/
 
-    public istring console_layout = "simple";
+    public istring console_layout;
 
     /***************************************************************************
 
@@ -167,7 +167,7 @@ class Config
 
     ***************************************************************************/
 
-    public istring file_layout = "date";
+    public istring file_layout;
 
     /***************************************************************************
 ```

First sentence of relnotes changed because it was wrong, `configureLoggers` wasn't changed (at least not the public ones).

File: src/ocean/util/log/Config.d [hunk 2]
```diff
@@ -447,8 +453,11 @@ public void configureLoggers ( Source = ConfigParser, FileLayout = LayoutDate,
         m_config = an instance of the MetaConfig class
         file_appender = delegate which returns appender instances to write to
                         a file
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throw on error.
+                          By default, wraps `ocean.util.log.Config.newLayout`
         use_insert_appender = true if the InsertConsole appender should be used
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: the `makeLayout` parameter only exists for the function that overloads `ConfigureOldLoggers()`

File: src/ocean/util/log/Config.d [hunk 3]
```diff
@@ -457,6 +466,18 @@ public void configureOldLoggers (
...
+    // link failures in D1
+    configureOldLoggers(config, m_config, file_appender,
+        (cstring v) { return newLayout(v); }, use_insert_appender);
+}
+
+/// Ditto
+public void configureOldLoggers (
+    ClassIterator!(Config, ConfigParser) config, MetaConfig m_config,
+    Appender delegate ( istring file, Layout layout ) file_appender,
+    Layout delegate (cstring) makeLayout, bool use_insert_appender = false)
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: The order of the last two parameters does not seem to be correct for the overloading function. I think it should be the other way around (`use_insert_appender` should go first)

**Review by daniel-zullo — COMMENTED** [review]
I'm about to test this patch in one application. Anyway I've done a couple of suggestions that you might want to address.

File: relnotes/makelayout.feature.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    ```
+    And calling `configureOldLoggers` should be done as such:
+    ```
+    void myConfigureLoggers (
+        ClassIterator!(Config, ConfigParser) config,
+        MetaConfig m_config,
+        Appender delegate ( istring file, Layout layout ) file_appender,
+        bool use_insert_appender = false)
+    {
+        scope makeLayout = (cstring name)
```
Comment (daniel-zullo) [inline relnotes/makelayout.feature.md]: I don't think this is correct. What would be the type of `makeLayout`? It might work for D2 though

File: relnotes/makelayout.feature.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            if (name == "aquatic")
+                return new AquaticLayout;
+            if (name == "submarine")
+                return new SubmarineLayout;
+            return ocean.util.log.Config.newLayout(name);
+        }
+        ocean.util.log.Config.configureOldLoggers(config, m_config,
+            file_appender, use_insert_appender, makeLayout);
+    }
+    ```
```
Comment (daniel-zullo) [inline relnotes/makelayout.feature.md]: Wouldn't it be more convenient to add the example in a unit test rather than in the release notes? Or maybe in both places

File: src/ocean/util/log/Config.d (hunk 3 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: It cannot go first because it has a default value

File: src/ocean/util/log/Config.d (hunk 2 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: This comment is on L450 / 459, and I'm not sure what the problem is ? This is the documentation for `configureOldLoggers`.

File: relnotes/makelayout.feature.md (hunk 4 above)
Comment (mathias-lang-sociomantic) [reply relnotes/makelayout.feature.md]: `scope Layout delegate (cstring)`

D1 is capable of inferring it (there are some bugs, but usually it works).
Maybe it's surprising because we usually use it with `class`es, but it works just the same with delegates.
Would writing the complete type make it more clear ?

File: src/ocean/util/log/Config.d (hunk 2 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: I just wanted to point out that `makeLayout` only exists in one of the `configureOldLoggers()` and it might not be right to have the same documentation for both methods when the signatures are different.

Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: Ah I see. I don't know if we have any guidelines about it. It's the style used in Phobos AFAIR, and DDOC just seem to handle it correctly.

File: relnotes/makelayout.feature.md (hunk 4 above)
Comment (daniel-zullo) [reply relnotes/makelayout.feature.md]: > D1 is capable of inferring it (there are some bugs, but usually it works).

Then I am afraid it will be more reliable to define the complete type (and clear).
The complete type might look a bit long though
```D
scope Layout delegate(cstring) makeLayout = (cstring name)
...
```

File: src/ocean/util/log/Config.d (hunk 3 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: Right, i wonder if it makes sense and it's doable to provide a default value for `makeLayout`
(I am not a big fan of switching the order of the parameters when overloading a function)

**Comment by mathias-lang-sociomantic** [comment]
So we tried it with @daniel-zullo-sociomantic and hit a little snag: In order to work reliably, it also needs to change `LogExt` to use the user defined `makeLayout`.

I was wondering how to better handle this: By introducting something specific to `makeLayout` (e.g. an extra parameter to`DaemonApp`'s ctor), something a bit more generic towards `LogExt` (e.g. a ctor that accepts a `LogExt` / `protected void setupLogExtension` function), or something else ?
CC @gavin-norman-sociomantic

File: src/ocean/util/log/Config.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/util/log/Config.d]: This is one reason I don't like `/// ditto` generally -- it's usually an over-simplification. (There must be _some_ difference between the two things that's worth explaining, otherwise why have them both?)

**Comment by gavin-norman-sociomantic** [comment]
A new field in [`DaemonApp.OptionalSettings`](https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/util/app/DaemonApp.d#L219) perhaps?

File: src/ocean/util/log/Config.d (hunk 2 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: > otherwise why have them both

Originally I went with a single overload with 2 default parameter, but then the linker said no (IOW, I hit a DMD bug and it's a workaround).

File: src/ocean/util/log/Config.d (hunk 3 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: See my comment [here](https://github.com/sociomantic-tsunami/ocean/pull/34#discussion_r106218478). I tried the default argument approach but hit a bug.
It's mentioned at L469 - 470:
```diff
+    // Using a dg literal as default argument in parameter list causes
+    // link failures in D1
```

**Comment by mathias-lang-sociomantic** [comment]
> A new field in DaemonApp.OptionalSettings perhaps?

Done. Will test with @daniel-zullo-sociomantic tomorrow.

**Comment by daniel-zullo** [comment]
The current patch is running smoothly in one application. You could check the reference to the workaround done in the application side to make it work (it's a hacky version). 
So I'd say now we can proceed with the review.

File: src/ocean/util/log/Config.d [hunk 6]
```diff
@@ -596,6 +621,9 @@ private void configureLoggers
         console_enabled = `true` if a console appender should be added (by
                           calling `console_enabled`).
         syslog_enabled  = `true` if a syslog appender should be added.
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throw on error.
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: throw**s**?

File: src/ocean/util/log/Config.d [hunk 7]
```diff
@@ -633,8 +662,8 @@ public void configureLogger
 
     if (console_enabled)
     {
-        Layout console_log_layout = (settings.console_layout.length)
-            ? newLayout(settings.console_layout)
+       Layout console_log_layout = (settings.console_layout.length)
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: The indentation seems to be wrong in this line

File: src/ocean/util/log/Config.d [hunk 8]
```diff
@@ -502,6 +523,8 @@ public void configureOldLoggers (
         console_appender = Delegate which returns an Appender suitable to use
                            as console appender. Might not be called if console
                            writing is disabled.
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throw on error.
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: ditto if my similar comment is taken

File: src/ocean/util/log/Config.d [hunk 9]
```diff
@@ -447,8 +453,11 @@ public void configureLoggers ( Source = ConfigParser, FileLayout = LayoutDate,
         m_config = an instance of the MetaConfig class
         file_appender = delegate which returns appender instances to write to
                         a file
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throw on error.
+                          By default, wraps `ocean.util.log.Config.newLayout`
         use_insert_appender = true if the InsertConsole appender should be used
-                              (needed when using the AppStatus module)
+                              (needed when using the AppStatus module).
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: the change in this line might not be related.

File: src/ocean/util/app/ext/LogExt.d [hunk 10]
```diff
@@ -199,4 +215,26 @@ class LogExt : IConfigExtExtension
         // Unused
         return files;
     }
+
+
+    /***************************************************************************
+
+        Default for layout_maker. DMD complains. because `LogUtil.newLayout`
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: You might want to delete the dot `.` appended to `complains`

File: src/ocean/util/app/ext/LogExt.d [hunk 11]
```diff
@@ -79,22 +79,38 @@ class LogExt : IConfigExtExtension
...
+    this ( bool use_insert_appender )
     {
-        this.use_insert_appender = use_insert_appender;
+        this(null, use_insert_appender);
     }
 
 
+    /// Ditto
+    public this ( Appender.Layout delegate (cstring name) make_layout = null,
+                 bool use_insert_appender = false)
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: The code style of the ctor signature doesn't seem to be consistent. Either you keep space for open/closed parenthesis or you remove for both of them

File: src/ocean/util/app/ext/LogExt.d [hunk 12]
```diff
@@ -79,22 +79,38 @@ class LogExt : IConfigExtExtension
...
+
+
     /***************************************************************************
 
         Constructor.
 
         Params:
+            make_layout         = A delegate that instantiate an
+                                  `Appender.Layout` from a name. If null,
+                                   defaults to `ocean.util.Config: newLayout`.
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: Please check indentation. Also this documentation should be added to the constructor below and not here. Again the `ditto` situation...

File: src/ocean/util/app/DaemonApp.d [hunk 13]
```diff
@@ -287,6 +288,9 @@ public abstract class DaemonApp : Application,
         ***********************************************************************/
 
         int reopen_signal = SIGHUP;
+
+        /// Delegate for LogExt that instantiate a `Appender.Layout` from a name
```
Comment (daniel-zullo) [inline src/ocean/util/app/DaemonApp.d]: instantiate**s**?

File: relnotes/makelayout.feature.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
+* `ocean.util.log.Config`
+
+    The `configureLogger` and `configureOldLoggers` functions
+    have gained an additional `makeLayout` parameter.
+    This parameter is used to instantiate a `Layout` from a name, allowing
+    users to have custom layout usable from their config files.
```
Comment (daniel-zullo) [inline relnotes/makelayout.feature.md]: either `a custom layout` or `custom layouts`

File: relnotes/makelayout.feature.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            if (name == "submarine")
+                return new SubmarineLayout;
+            return ocean.util.log.Config.newLayout(name);
+        }
+        ocean.util.log.Config.configureOldLoggers(config, m_config,
+            file_appender, use_insert_appender, makeLayout);
+    }
+    ```
+    Note that the previous example is kept short for simplicity. `makeLayout` gets the raw name,
+    and as such should make sure to make the comparison as loose as wanted (by lowercasing,
```
Comment (daniel-zullo) [inline relnotes/makelayout.feature.md]: Maybe replace `make sure` to `ensure`.

File: src/ocean/util/app/ext/LogExt.d [hunk 16]
```diff
@@ -79,22 +79,38 @@ class LogExt : IConfigExtExtension
...
             use_insert_appender = true if the InsertConsole appender should be
                                   used (needed when using the AppStatus module)
 
     ***************************************************************************/
 
-    this ( bool use_insert_appender = false )
+    this ( bool use_insert_appender )
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d:96]: Why the default value is removed here?

File: src/ocean/util/app/ext/LogExt.d [hunk 17]
```diff
@@ -79,22 +79,38 @@ class LogExt : IConfigExtExtension
...
+        this(null, use_insert_appender);
     }
 
 
+    /// Ditto
+    public this ( Appender.Layout delegate (cstring name) make_layout = null,
+                 bool use_insert_appender = false)
+    {
+        this.layout_maker = make_layout !is null ? make_layout
+            : &this.makeLayoutDefault;
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: What about doing it the other way around:
```D
    this.layout_maker = make_layout is null ? &this.makeLayoutDefault
        : make_layout;
```

File: src/ocean/util/log/Config.d [hunk 18]
```diff
@@ -447,8 +453,11 @@ public void configureLoggers ( Source = ConfigParser, FileLayout = LayoutDate,
         m_config = an instance of the MetaConfig class
         file_appender = delegate which returns appender instances to write to
                         a file
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throw on error.
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: throw**s**?

File: src/ocean/util/app/ext/LogExt.d [hunk 19]
```diff
@@ -199,4 +215,26 @@ class LogExt : IConfigExtExtension
...
     }
+
+
+    /***************************************************************************
+
+        Default for layout_maker. DMD complains. because `LogUtil.newLayout`
+        is a function, now a delegate.
+
+        Params:
+            name = Name of the Layout to instantiate.
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: `Name` is Capitalized

File: src/ocean/util/app/ext/LogExt.d [hunk 20]
```diff
@@ -199,4 +215,26 @@ class LogExt : IConfigExtExtension
         // Unused
         return files;
     }
+
+
+    /***************************************************************************
+
+        Default for layout_maker. DMD complains. because `LogUtil.newLayout`
+        is a function, now a delegate.
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: `now a delegate`? what do you mean here?

**Review by daniel-zullo — COMMENTED** [review]
I've finished my review round. I just have some minor comments and a couple of questions. We are also still discussing a couple of issues related to the design of this feature.

File: src/ocean/util/app/ext/LogExt.d (hunk 16 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/LogExt.d:96]: Because there are 2 ways to call this ctor:
`this()` and `this(bool)`.
I don't want a breaking change, so `this(bool)` needs to still work, but `this()` now calls the other ctor (which is compatible).

File: src/ocean/util/app/ext/LogExt.d (hunk 20 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/LogExt.d]: Typo, should be `not`.

File: src/ocean/util/app/ext/LogExt.d (hunk 16 above)
Comment (daniel-zullo) [reply src/ocean/util/app/ext/LogExt.d:96]: Got it, thanks for the clarification.

File: src/ocean/util/log/Config.d (hunk 7 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: It was wrong before actually.

File: relnotes/makelayout.feature.md (hunk 15 above)
Comment (mathias-lang-sociomantic) [reply relnotes/makelayout.feature.md]: Is the current wording incorrect ? Ensure sounds more like an absolute requirement, which this is a pretty loose area.

**Comment by mathias-lang-sociomantic** [comment]
Updated accorded to @daniel-zullo-sociomantic 's (except the ones I answered to).

File: src/ocean/util/app/ext/LogExt.d [hunk 21]
```diff
@@ -79,22 +79,39 @@ class LogExt : IConfigExtExtension
...
+        this(null, use_insert_appender);
     }
 
 
+    /// Ditto
+    public this ( Appender.Layout delegate (cstring name) make_layout = null,
+                 bool use_insert_appender = false )
+    {
+        this.layout_maker = make_layout is null ? &this.makeLayoutDefault
+            : make_layout
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: semicolon is missing at the end of this line

File: src/ocean/util/log/Config.d (hunk 7 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: mmm...I don't think so. You might need to add one space before `Layout`

**Comment by mathias-lang-sociomantic** [comment]
Fixed and green

File: src/ocean/util/app/ext/LogExt.d (hunk 21 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/LogExt.d]: Forgot to mention, this was addressed.

File: src/ocean/util/log/Config.d (hunk 7 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: it seems this comment was not addressed, was it?

File: src/ocean/util/log/Config.d (hunk 8 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: also here you might want to append `s` to `throw`

File: src/ocean/util/log/Config.d (hunk 18 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: and this one?

File: src/ocean/util/log/Config.d [hunk 22]
```diff
@@ -427,11 +429,15 @@ public void configureLoggers ( Source = ConfigParser, FileLayout = LayoutDate,
 
     enable_loose_parsing(loose);
 
+    scope Layout delegate(cstring) makeLayout
+                     = (cstring v) { return newLayout(v); };
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: it seems the indentation is not right

File: src/ocean/util/app/ext/LogExt.d [hunk 23]
```diff
@@ -79,22 +79,39 @@ class LogExt : IConfigExtExtension
...
+    /// If set, will be used by `makeLayout` to map a name to a `Layout`
+    private Appender.Layout delegate (cstring name) layout_maker;
+
+
     /***************************************************************************
 
         Constructor.
 
         Params:
+            make_layout         = A delegate that instantiate an
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: instantiate**s**

**Review by daniel-zullo — COMMENTED** [review]
I've made just a couple of cosmetic comments and LGTM. Also I've tested with one application and it works fine. Anyway it might be worth to have a full review from someone with more experience with the `Logger`

**Comment by mathias-lang-sociomantic** [comment]
Addressed comments by @daniel-zullo-sociomantic

File: src/ocean/util/log/Config.d (hunk 7 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: > mmm...I don't think so. You might need to add one space before `Layout`

One level of indent = 4 spaces. There were 5 before. I do not do this manually, my editor caught it.

**Comment by mathias-lang-sociomantic** [comment]
Pingitty ping

**Comment by nemanja-boric-sociomantic** [comment]
I'll look into it, but @gavin-norman-sociomantic has more experience with Config parsers.

File: src/ocean/util/log/Config.d [hunk 24]
```diff
@@ -476,10 +497,35 @@ public void configureOldLoggers (
                        { return console_appender_fn(use_insert_appender, l); };
 
     configureLoggers!(Logger, ConfigParser, LayoutDate, LayoutSimple)
-        (config, m_config, lookup, file_appender, appender_dg);
+        (config, m_config, lookup, file_appender, appender_dg, makeLayout);
 }
 
 
+unittest
```
Comment (mathias-baumann-sociomantic) [inline src/ocean/util/log/Config.d]: Should this maybe be a documented unittest, serving as example as well?

File: src/ocean/util/app/ext/LogExt.d [hunk 25]
```diff
@@ -79,22 +79,39 @@ class LogExt : IConfigExtExtension
...
+    this ( bool use_insert_appender )
     {
-        this.use_insert_appender = use_insert_appender;
+        this(null, use_insert_appender);
     }
 
 
+    /// Ditto
+    public this ( Appender.Layout delegate (cstring name) make_layout = null,
+                 bool use_insert_appender = false )
```
Comment (mathias-baumann-sociomantic) [inline src/ocean/util/app/ext/LogExt.d]: I think you need one whitespace more here

**Review by mathias-baumann-sociomantic — CHANGES_REQUESTED** [review]

**Comment by mathias-lang-sociomantic** [comment]
Addressed @mathias-baumann-sociomantic 's comments

**Review by mathias-baumann-sociomantic — APPROVED** [review]

File: relnotes/makelayout.feature.md (hunk 15 above)
Comment (gavin-norman-sociomantic) [reply relnotes/makelayout.feature.md]: "and as such may make the comparison as loose as wanted" is clearer.

File: src/ocean/util/log/Config.d [hunk 26]
```diff
@@ -315,7 +315,9 @@ public Layout newLayout ( istring layout_str )
             break;
 
         default:
-            throw new Exception("Invalid layout requested : " ~ layout_str);
+            istring msg = "Invalid layout requested : ";
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/log/Config.d]: Why this code formatting change? Adding file and line is unrelated to this commit, right?

File: src/ocean/util/log/Config.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/util/log/Config.d]: I'd go with duplicating the doc (don't use `/// ditto`) and mention the reason for the two overloads in the doc for the second (i.e. move it out of the method body and into its "header" doc).

File: relnotes/makelayout.feature.md [hunk 27]
```diff
@@ -0,0 +1,37 @@
+* `ocean.util.log.Config`
+
+    The `configureLogger` and `configureOldLoggers` functions
```
Comment (gavin-norman-sociomantic) [inline relnotes/makelayout.feature.md]: The way this is described "have gained" makes this sound like a breaking change. It should be made clear that these are overloads, not replacements.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Apart from a few minor points, LGTM. (Sorry for the delay.)

File: src/ocean/util/log/Config.d (hunk 26 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: The change was "necessary" because `istring ~ cstring` yields `cstring` (stupid, I know). Thus `throw new Exception("..." ~ layout_str);` stopped to compile after I changed `layout_str` to `cstring`.
The options, were to add a `cast` (unnecessary and prone to refactoring errors), use `format` (adding a new dependency seemed overkill), or just break down the statement. And since I was changing those line, I figured I might just add `__FILE__` and `__LINE__`.

Comment (gavin-norman-sociomantic) [reply src/ocean/util/log/Config.d]: Ha gotcha. Pretty please split the file & line change into another commit? Just a minor thing, but I think that'll make the other change here -- which is actually connected to this commit -- easier to spot.

**Comment by mathias-lang-sociomantic** [comment]
Updated:

```diff
diff --git a/relnotes/makelayout.feature.md b/relnotes/makelayout.feature.md
index e50d128..903a01f 100644
--- a/relnotes/makelayout.feature.md
+++ b/relnotes/makelayout.feature.md
@@ -1,7 +1,7 @@
 * `ocean.util.log.Config`
 
-    The `configureLogger` and `configureOldLoggers` functions
-    have gained an additional `makeLayout` parameter.
+    An extra overload of the `configureLogger` and `configureOldLoggers` functions
+    were introduced, with an additional `makeLayout` parameter.
     This parameter is used to instantiate a `Layout` from a name, allowing
     users to have custom layouts usable from their config files.
     For example, a config entry would be:
@@ -33,8 +33,8 @@
     }
     ```
     Note that the previous example is kept short for simplicity. `makeLayout` gets the raw name,
-    and as such should make sure to make the comparison as loose as wanted (by lowercasing,
-    removing `"layout"` prefix/suffix, etc...).
+    and as such may make the comparison as loose as wanted
+    (by lowercasing, removing `"layout"` prefix/suffix, etc...).
 
 * `ocean.util.app.LogExt : LogExt`
```

```diff
diff --git a/src/ocean/util/log/Config.d b/src/ocean/util/log/Config.d
index 8a861e3..8c7659a 100644
--- a/src/ocean/util/log/Config.d
+++ b/src/ocean/util/log/Config.d
@@ -453,9 +453,6 @@ public void configureLoggers ( Source = ConfigParser, FileLayout = LayoutDate,
         m_config = an instance of the MetaConfig class
         file_appender = delegate which returns appender instances to write to
                         a file
-        makeLayout      = A delegate that returns a `Layout` instance from
-                          a name, or throws on error.
-                          By default, wraps `ocean.util.log.Config.newLayout`
         use_insert_appender = true if the InsertConsole appender should be used
                               (needed when using the AppStatus module)
 
@@ -466,13 +463,34 @@ public void configureOldLoggers (
     Appender delegate ( istring file, Layout layout ) file_appender,
     bool use_insert_appender = false)
 {
-    // Using a dg literal as default argument in parameter list causes
-    // link failures in D1
     configureOldLoggers(config, m_config, file_appender,
         (cstring v) { return newLayout(v); }, use_insert_appender);
 }
 
-/// Ditto
+
+/*******************************************************************************
+
+    Sets up logging configuration for `ocean.util.log.Log`
+
+    Calls the provided `file_appender` delegate once per log being configured
+    and passes the returned `Appender` to the `Logger.add` method.
+
+    This is an extra overload because using a delegate literal as a parameter's
+    default argument causes linker error in D1.
+
+    Params:
+        config   = an instance of an class iterator for Config
+        m_config = an instance of the MetaConfig class
+        file_appender = delegate which returns appender instances to write to
+                        a file
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throws on error.
+                          By default, wraps `ocean.util.log.Config.newLayout`
+        use_insert_appender = true if the InsertConsole appender should be used
+                              (needed when using the AppStatus module)
+
+*******************************************************************************/
+
 public void configureOldLoggers (
     ClassIterator!(Config, ConfigParser) config, MetaConfig m_config,
     Appender delegate ( istring file, Layout layout ) file_appender,
```

**Comment by mathias-lang-sociomantic** [comment]
Updated:

```diff
diff --git a/src/ocean/util/log/Config.d b/src/ocean/util/log/Config.d
index 8c7659a..a816a1f 100644
--- a/src/ocean/util/log/Config.d
+++ b/src/ocean/util/log/Config.d
@@ -315,6 +315,8 @@ public Layout newLayout ( cstring layout_str )
             break;
 
         default:
+            // Has to be 2 statements because `istring ~ cstring`
+            // yields `cstring` instead of `istring`.
             istring msg = "Invalid layout requested : ";
             msg ~= layout_str;
             throw new Exception(msg, __FILE__, __LINE__);
```

+ split the `__FILE__` and `__LINE__` additions.

File: src/ocean/util/log/Config.d (hunk 26 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: Sure, done, and added a comment.

**Review by gavin-norman-sociomantic — APPROVED** [review]
LGTM! @daniel-zullo-sociomantic as the person who filed the issue, does this meet your approval too?

File: src/ocean/util/app/ext/LogExt.d [hunk 28]
```diff
@@ -79,22 +79,39 @@ class LogExt : IConfigExtExtension
...
 
-    this ( bool use_insert_appender = false )
+    this ( bool use_insert_appender )
     {
-        this.use_insert_appender = use_insert_appender;
+        this(null, use_insert_appender);
     }
 
 
+    /// Ditto
```
Comment (daniel-zullo) [inline src/ocean/util/app/ext/LogExt.d]: You might not use Ditto here (to be consistent with other similar situations in other functions). So new documentation needs to be added to this constructor and the parameter `make_layout`  should be removed from the constructor above

File: src/ocean/util/log/Config.d [hunk 29]
```diff
@@ -457,6 +465,39 @@ public void configureOldLoggers (
...
+    Params:
+        config   = an instance of an class iterator for Config
+        m_config = an instance of the MetaConfig class
+        file_appender = delegate which returns appender instances to write to
+                        a file
+        makeLayout      = A delegate that returns a `Layout` instance from
+                          a name, or throws on error.
+                          By default, wraps `ocean.util.log.Config.newLayout`
+        use_insert_appender = true if the InsertConsole appender should be used
+                              (needed when using the AppStatus module)
```
Comment (daniel-zullo) [inline src/ocean/util/log/Config.d]: The format of the documentation needs to be well formatted.

**Review by daniel-zullo — COMMENTED** [review]
Apart from my 2 minor comments LGTM

File: src/ocean/util/log/Config.d (hunk 29 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/log/Config.d]: It is literaly a copy/paste of what's before, just adding a parameter to the list. I don't see how it's badly formatted ?

**Comment by mathias-lang-sociomantic** [comment]
Updated:

```diff
diff --git a/src/ocean/util/app/ext/LogExt.d b/src/ocean/util/app/ext/LogExt.d
index e8851af..ce27ffc 100644
--- a/src/ocean/util/app/ext/LogExt.d
+++ b/src/ocean/util/app/ext/LogExt.d
@@ -88,9 +88,6 @@ class LogExt : IConfigExtExtension
         Constructor.
 
         Params:
-            make_layout         = A delegate that instantiates an
-                                  `Appender.Layout` from a name. If null,
-                                  defaults to `ocean.util.Config: newLayout`.
             use_insert_appender = true if the InsertConsole appender should be
                                   used (needed when using the AppStatus module)
 
@@ -102,7 +99,19 @@ class LogExt : IConfigExtExtension
     }
 
 
-    /// Ditto
+    /***************************************************************************
+
+        Constructor.
+
+        Params:
+            make_layout         = A delegate that instantiates an
+                                  `Appender.Layout` from a name. If null,
+                                  defaults to `ocean.util.Config: newLayout`.
+            use_insert_appender = true if the InsertConsole appender should be
+                                  used (needed when using the AppStatus module)
+
+    ***************************************************************************/
+
     public this ( Appender.Layout delegate (cstring name) make_layout = null,
                   bool use_insert_appender = false )
     {
```

File: src/ocean/util/log/Config.d (hunk 29 above)
Comment (daniel-zullo) [reply src/ocean/util/log/Config.d]: I think `makeLayout` does not have the right format and that is a new parameter (the spaces between `makeLayout`  and `=` are not consistent to the other parameters)

**Comment by mathias-lang-sociomantic** [comment]
Updated:

```diff
diff --git a/src/ocean/util/log/Config.d b/src/ocean/util/log/Config.d
index a816a1f..04cfa1a 100644
--- a/src/ocean/util/log/Config.d
+++ b/src/ocean/util/log/Config.d
@@ -485,9 +485,9 @@ public void configureOldLoggers (
         m_config = an instance of the MetaConfig class
         file_appender = delegate which returns appender instances to write to
                         a file
-        makeLayout      = A delegate that returns a `Layout` instance from
-                          a name, or throws on error.
-                          By default, wraps `ocean.util.log.Config.newLayout`
+        makeLayout = A delegate that returns a `Layout` instance from
+                     a name, or throws on error.
+                     By default, wraps `ocean.util.log.Config.newLayout`
         use_insert_appender = true if the InsertConsole appender should be used
                               (needed when using the AppStatus module)
 ```

**Review by daniel-zullo — APPROVED** [review]
LGTM, many thanks!

**Comment by mathias-lang-sociomantic** [comment]
Thanks everyone for the review :)
